File: python/difference-of-squares/difference_of_squares.py
"""
Exercism.io assignment - difference_of_squares.py
Written by cmdellinger
    
Usage:
    difference_of_squares.py <number> (--sosq|--sqos|--diff)
    
Options:
    --sosq  Sum of squares of natural numbers up to input number
    --sqos  Square of sums of natural numbers up to input number
    --diff  Difference between the sum of squares and the square of sums
    
Find the sum of the squares, the square of the sum, or the difference between the two
for the sequence up to the number provided.
"""

## ------
## Exercism.io solution
## ------

def square_of_sum(number = 0): # -> integer
    ''' returns the square of the sums of natural numbers up to input number '''
    total = 0
    for integer in xrange(number + 1):
        total += integer
    return total**2

def sum_of_squares(number = 0): # -> integer
    ''' returns the sum of the squares of natural numbers up to input number '''
    total = 0
    for integer in xrange(number + 1):
        total += integer**2
    return total

def difference(number = 0): # -> integer
    ''' returns the difference between the sum of squares and the square of the sum '''
    # square_of_sum is always greater than sum_of_squares
    return square_of_sum(number) - sum_of_squares(number)

## ------
## Command-line implementation
## ------

if __name__ == '__main__':
    from docopt import docopt
    number = int(docopt(__doc__)['<number>'])
    #if docopt(__doc__)['--sosq'] or docopt(__doc__)['--sqos'] or docopt(__doc__)['--diff']:
    if docopt(__doc__)['--sosq']:
        print "Sum of Squares: %d" % sum_of_squares(number)
    if docopt(__doc__)['--sqos']:
        print "Square of Sum: %d" % sum_of_squares(number)
    if docopt(__doc__)['--diff']:
        print "Difference: %d" % difference(number)

File: ruby/difference-of-squares/difference_of_squares.rb
# @title difference_of_squares
# @author cmdellinger

class Squares
  # Creates instance range sequence to be used in calculations.
  #
  # @param input_number [Integer] maximum of sequence from 1 to input_number.
  def initialize(input_number)
    @number_range = (1..input_number)
  end

  # Squares the sum of integers in the sequence of instance range.
  # Sequence range: 1 to input_number.
  #
  # @return [Integer] sum of the squares in object's range.
  def sum_of_squares
    @number_range.inject{|memo, n| memo + n**2}
  end
  
  # Squares the sum of integers in the sequence of instance range.
  # Sequence range: 1 to input_number.
  #
  # @return [Integer] sum of the squares in object's range.
  def square_of_sum
    @number_range.sum**2
  end
  
  # Calculates the difference between the square of sums and the sum of squares
  # for the instance range.
  #
  # @return [Integer] difference between the square of sums and the sum of squares.
  def difference
    # square of sums will always be bigger
    square_of_sum - sum_of_squares
  end
end

# BookKeeping Version
module BookKeeping
  # the version of the tests that are being run
  VERSION = 4 # Where the version number matches the one in the test.
end
